### PR TITLE
Refactored code for readability

### PR DESCRIPTION
**Pull request by sateffen**
Hey,

this PR is basically a refactoring of the code-basis for ts-nats. The changes are only code-style changes, more or less, and I hope to make the code-base simpler for future contributors. I didn't change the logic of the code. Not all changes might be wanted, so let's talk about them, so I can learn what's needed.

The changelog:

* Added/removed spacing and empty lines where needed
* Added/removed semicolons where needed
* Changed the copyright year to 2019 in all edited files
* Added missing return-type definitions
* Added code comments at 1-2 points
* Unified import style
    * Some imports used `import x = require('...')`, other `import x from '...'`, now it's the second in all files
* Changed `let` to `const` in all places where possible
    * This prevents unintentional behavior straight up while developing, and prevents possible problems when working on future features
* Replaced a lot of single character variable names with multi-character ones
    * Reading the code with single character variable names is pretty hard, and makes the code harder to grasp. Helpful names make code more approachable
    * Some short names are not optimal, for example `catch(ex)`, where `ex` implies an exception, not an error. Errors are runtime-errors, that occur in own code, exceptions are problems/bugs in the JS-engine itself, so error is the better fitting one (in my opinion)
    * Some names like `fs` imply a certain content (for `fs` it's the filesystem), so I've tried to choose an appropriate one
* Refactored some forEach-loops to map-loops to make the code more expressive

### CODE REVIEW

**Comment by sateffen** [comment]
The current build problems are related to [this](https://github.com/highlightjs/highlight.js/issues/1984) and [this](https://github.com/highlightjs/highlight.js/issues/1986) issue. As soon as the issue with highlight.js is resolved the builds should work.

File: src/const.ts [hunk 1]
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright 2018 The NATS Authors
+ * Copyright 2019 The NATS Authors
```
Comment (aricart) [inline src/const.ts]: must be 2018-19

File: src/const.ts [hunk 2]
```diff
@@ -26,7 +27,7 @@ export const DEFAULT_MAX_RECONNECT_ATTEMPTS = 10;
 export const DEFAULT_PING_INTERVAL = 2 * 60 * 1000; // 2 minutes
 export const DEFAULT_MAX_PING_OUT = 2;
 
-
+// Line handling
```
Comment (aricart) [inline src/const.ts]: OK

File: src/databuffer.ts [hunk 3]
```diff
@@ -30,11 +30,11 @@ export class DataBuffer {
     buffers: Buffer[] = [];
     byteLength: number = 0;
 
-
     pack(): void {
         if (this.buffers.length > 1) {
-            let v = Buffer.concat(this.buffers, this.byteLength);
-            this.buffers = [v];
+            const compactedBuffer = Buffer.concat(this.buffers, this.byteLength);
+
```
Comment (aricart) [inline src/databuffer.ts]: Remove extra line

File: src/databuffer.ts [hunk 4]
```diff
@@ -45,16 +45,20 @@ export class DataBuffer {
 
         if (this.byteLength >= n) {
             this.pack();
-            let v = this.buffers.pop();
-            if (v) {
-                let d = v.slice(0, n);
+            const dataBuffer = this.buffers.pop();
```
Comment (aricart) [inline src/databuffer.ts]: This is a local variable that spans very few lines. The new name makes it seem as an important variable or object, please keep the previous name.

File: src/error.ts [hunk 5]
```diff
@@ -14,7 +14,7 @@
  *
  */
 
-import util = require('util');
+import * as util from 'util';
```
Comment (aricart) [inline src/error.ts]: Yes!

File: src/error.ts [hunk 6]
```diff
@@ -87,20 +86,21 @@ export class Messages {
         this.messages[ErrorCode.SUB_TIMEOUT] = 'Subscription timed out.';
     }
 
-    static getMessage(s: string): string {
-        return Messages.messages.getMessage(s);
+    static getMessage(errorCode: string): string {
```
Comment (aricart) [inline src/error.ts]: name it `code`

File: src/error.ts [hunk 7]
```diff
@@ -87,20 +86,21 @@ export class Messages {
...
+        return Messages.messages.getMessage(errorCode);
     }
 
-    getMessage(s: string): string {
-        let v = this.messages[s];
-        if (!v) {
-            v = s;
+    getMessage(errorCode: string): string {
```
Comment (aricart) [inline src/error.ts]: name it `code`

File: src/error.ts [hunk 8]
```diff
@@ -87,20 +86,21 @@ export class Messages {
...
+        let errorMessage = this.messages[errorCode];
+
```
Comment (aricart) [inline src/error.ts]: remove extra white lines on this little method.

File: src/error.ts [hunk 9]
```diff
@@ -117,7 +117,7 @@ export class NatsError implements Error {
      */
     constructor(message: string, code: string, chainedError?: Error) {
         Error.captureStackTrace(this, this.constructor);
-        this.name = "NatsError";
+        this.name = 'NatsError';
```
Comment (aricart) [inline src/error.ts]: Restore string quotes to their original value.

File: src/error.ts [hunk 10]
```diff
@@ -126,7 +126,8 @@ export class NatsError implements Error {
     }
 
     static errorForCode(code: string, chainedError?: Error): NatsError {
-        let m = Messages.getMessage(code);
-        return new NatsError(m, code, chainedError);
+        const errorMessage = Messages.getMessage(code);
+
```
Comment (aricart) [inline src/error.ts]: Remove extra line

File: src/muxsubscriptions.ts [hunk 11]
```diff
@@ -29,60 +28,69 @@ export class MuxSubscriptions {
...
         }
     }
 
-    getToken(m?: Msg): string | null {
-        let s = "";
-        if (m) {
-            s = m.subject || "";
+    getToken(message?: Msg): string | null {
+        let subject = '';
+
```
Comment (aricart) [inline src/muxsubscriptions.ts]: Remove extra empty lines on small method

File: src/nats.ts [hunk 12]
```diff
@@ -547,10 +550,10 @@ export class Subscription {
     /**
      * @hidden
      */
-    static cancelTimeout(s: Sub | null): void {
-        if (s && s.timeout) {
-            clearTimeout(s.timeout);
-            delete s.timeout;
+    static cancelTimeout(subscription: Sub | null): void {
```
Comment (aricart) [inline src/nats.ts]: favor shorter names, like `sub`, `msg`, etc.

File: src/protocolhandler.ts [hunk 13]
```diff
@@ -27,30 +27,28 @@ import {
...
-import {TCPTransport} from "./tcptransport";
-import {Subscriptions} from "./subscriptions";
-import {DataBuffer} from "./databuffer";
-import {MsgBuffer} from "./messagebuffer";
-import {settle} from "./util";
-import fs = require('fs');
-import url = require('url');
-import nkeys = require('ts-nkeys');
+} from './nats';
+import { MuxSubscriptions } from './muxsubscriptions';
```
Comment (aricart) [inline src/protocolhandler.ts]: remove whitespace after and before the braces

File: src/protocolhandler.ts [hunk 14]
```diff
@@ -196,21 +191,20 @@ export class ProtocolHandler extends EventEmitter {
...
         this.draining = true;
 
-        let subs = this.subscriptions.all();
-        let promises: Promise<SubEvent>[] = [];
-        subs.forEach((sub) => {
-            let p = this.drainSubscription(sub.sid);
-            promises.push(p);
-        });
+        const subs = this.subscriptions.all();
+        let promises: Promise<SubEvent>[] = subs.map((sub) => this.drainSubscription(sub.sid));
```
Comment (aricart) [inline src/protocolhandler.ts]: please add back the initialization - now if for some reason the code changes and the map doesn't have that key, the iteration will fail with an NPE

File: src/protocolhandler.ts [hunk 15]
```diff
@@ -225,40 +219,49 @@ export class ProtocolHandler extends EventEmitter {
...
             proto = `PUB ${subject} ${reply} ${len}`;
         } else {
             proto = `PUB ${subject} ${len}`;
         }
+
         this.sendCommand(this.buildProtocolMessage(proto, data))
     }
 
-    subscribe(s: Sub): Subscription {
+    subscribe(subscription: Sub): Subscription {
```
Comment (aricart) [inline src/protocolhandler.ts]: `sub`

Comment (aricart) [inline src/protocolhandler.ts]: this is not a subscription but a descriptor- the actual subscription is created at the bottom.

File: src/protocolhandler.ts [hunk 16]
```diff
@@ -270,53 +273,60 @@ export class ProtocolHandler extends EventEmitter {
...
+
+            sub.max = max;
+            if (max === undefined || sub.received >= max) {
+                this.subscriptions.cancel(sub);
             }
         }
     }
 
-    request(r: Req): Request {
+    request(request: Req): Request {
```
Comment (aricart) [inline src/protocolhandler.ts]: this is not a request, but a descriptor for a request.

File: src/protocolhandler.ts [hunk 17]
```diff
@@ -562,14 +587,16 @@ export class ProtocolHandler extends EventEmitter {
...
-        this.subscriptions.all().forEach((s) => {
-            if (s.queue) {
-                cmds.push(`${SUB} ${s.subject} ${s.queue} ${s.sid} ${CR_LF}`);
-            } else {
-                cmds.push(`${SUB} ${s.subject} ${s.sid} ${CR_LF}`);
-            }
-        });
+
+        const cmds: string[] = this.subscriptions.all()
+            .map((subscription) => {
```
Comment (aricart) [inline src/protocolhandler.ts]: single letter variable name please

**Comment by aricart** [comment]
> The current build problems are related to [this](https://github.com/highlightjs/highlight.js/issues/1984) and [this](https://github.com/highlightjs/highlight.js/issues/1986) issue. As soon as the issue with highlight.js is resolved the builds should work.

Yes - when it works we can add it - node 6 is eol very soon now.

**Comment by sateffen** [comment]
Thank you for the feedback! The provided rules give me a good feeling about the code style for future PRs :)

Tomorrow I'll rework the PR according to your feedback and check all not commented changes based on that (here in Germany it's already late).

I want to ask about the reasoning behind `Changes from double-quote to single quote for strings will not be accepted`. The quoting (`'` or `"`) is pretty mixed throughout the code (example: protocolhandler.ts [L49](https://github.com/nats-io/ts-nats/blob/master/src/protocolhandler.ts#L49) vs [L64](https://github.com/nats-io/ts-nats/blob/master/src/protocolhandler.ts#L64)), so I thought about unifying this. Is there a special rule, or could I normalize them to double-quotes?

File: src/protocolhandler.ts (hunk 14 above)
Comment (sateffen) [reply src/protocolhandler.ts]: Could you elaborate on that? `this.subscriptions.all()` returns a `Sub[]`, not nullable, so `subs.map` [always exists and always returns an array](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/map). If `subs` is empty, it'll return an empty array, so `promises` is initialized in all cases.

`sub.sid` is not safeguarded in the previous version either, so maybe I'm missing something?

**Comment by aricart** [comment]
> Thank you for the feedback! The provided rules give me a good feeling about the code style for future PRs :)
> 
> Tomorrow I'll rework the PR according to your feedback and check all not commented changes based on that (here in Germany it's already late).
> 
> I want to ask about the reasoning behind `Changes from double-quote to single quote for strings will not be accepted`. The quoting (`'` or `"`) is pretty mixed throughout the code (example: protocolhandler.ts [L49](https://github.com/nats-io/ts-nats/blob/master/src/protocolhandler.ts#L49) vs [L64](https://github.com/nats-io/ts-nats/blob/master/src/protocolhandler.ts#L64)), so I thought about unifying this. Is there a special rule, or could I normalize them to double-quotes?

OK - compromise - let's do single quotes everywhere unless the string embeds a single quote in which case it will be an outer double quote.

**Comment by sateffen** [comment]
I'm going to split the changes in other PRs, so the git-history will stay clean. This PR won't get any future changes, so it's safe to close. Changes, that are not present in #38 , but are applicable will get their own PR.

**Comment by aricart** [comment]
@sateffen the other PR covered the changes to include, I am going to closing this one. Thank you for your contribution.

**Pull request closed without merging**